File: cpp/tests/test_batch_operations.cpp
#include <gtest/gtest.h>
#include "drawing/batch_operations.hpp"
#include <chrono>

using namespace drawing;

class BatchOperationsTest : public ::testing::Test {
protected:
    ObjectStorage storage;
    std::vector<ObjectID> test_objects;
    
    void SetUp() override {
        // Create test objects
        for (int i = 0; i < 10; ++i) {
            test_objects.push_back(storage.add_circle(i * 50, i * 50, 20));
            test_objects.push_back(storage.add_rectangle(i * 50, i * 50 + 100, 30, 40));
        }
    }
};

TEST_F(BatchOperationsTest, TranslateObjects) {
    // Get initial positions
    auto* first_circle = storage.get_circle(test_objects[0]);
    ASSERT_NE(first_circle, nullptr);
    float initial_x = first_circle->x;
    float initial_y = first_circle->y;
    
    // Translate all objects
    BatchOperations::translate_objects(storage, test_objects, 100, 50);
    
    // Check that objects moved
    EXPECT_FLOAT_EQ(first_circle->x, initial_x + 100);
    EXPECT_FLOAT_EQ(first_circle->y, initial_y + 50);
    
    // Check performance stats
    auto stats = BatchOperations::last_operation_stats;
    EXPECT_EQ(stats.objects_processed, test_objects.size());
    EXPECT_GT(stats.objects_per_second, 0);
}

TEST_F(BatchOperationsTest, ScaleObjects) {
    BatchOperations::scale_objects(storage, test_objects, 2.0f, 2.0f, Point(0, 0));
    
    // Check that first circle doubled in size and position
    auto* first_circle = storage.get_circle(test_objects[0]);
    EXPECT_FLOAT_EQ(first_circle->x, 0); // Was at 0, still at 0
    EXPECT_FLOAT_EQ(first_circle->radius, 40); // Was 20, now 40
    
    // Check second circle (was at 50,50)
    auto* second_circle = storage.get_circle(test_objects[2]);
    EXPECT_FLOAT_EQ(second_circle->x, 100); // 50 * 2
    EXPECT_FLOAT_EQ(second_circle->y, 100); // 50 * 2
}

TEST_F(BatchOperationsTest, RotateObjects) {
    // Create a simple line for rotation test
    auto line_id = storage.add_line(0, 0, 100, 0);
    std::vector<ObjectID> line_vec = {line_id};
    
    // Rotate 90 degrees around origin
    BatchOperations::rotate_objects(storage, line_vec, M_PI/2, Point(0, 0));
    
    auto* line = storage.get_line(line_id);
    ASSERT_NE(line, nullptr);
    
    // Should now be vertical
    EXPECT_NEAR(line->x1, 0, 0.001);
    EXPECT_NEAR(line->y1, 0, 0.001);
    EXPECT_NEAR(line->x2, 0, 0.001);
    EXPECT_NEAR(line->y2, 100, 0.001);
}

TEST_F(BatchOperationsTest, CalculateBoundingBox) {
    BoundingBox bbox = BatchOperations::calculate_bounding_box(storage, test_objects);
    
    // Should encompass all objects
    EXPECT_LE(bbox.min_x, 0);
    EXPECT_LE(bbox.min_y, 0);
    EXPECT_GE(bbox.max_x, 450); // Last circle at 450
    EXPECT_GE(bbox.max_y, 540); // Last rectangle at y=540
}

TEST_F(BatchOperationsTest, AlignObjectsLeft) {
    BatchOperations::align_objects_left(storage, test_objects);
    
    // Find leftmost edge
    float min_x = std::numeric_limits<float>::max();
    for (auto id : test_objects) {
        BoundingBox bbox = BatchOperations::get_object_bbox(storage, id);
        min_x = std::min(min_x, bbox.min_x);
    }
    
    // All objects should have their left edge at min_x
    for (auto id : test_objects) {
        BoundingBox bbox = BatchOperations::get_object_bbox(storage, id);
        EXPECT_FLOAT_EQ(bbox.min_x, min_x);
    }
}

TEST_F(BatchOperationsTest, CreateGrid) {
    ObjectStorage grid_storage;
    
    auto grid_objects = BatchOperations::create_grid(
        grid_storage, ObjectType::Circle,
        5, 5, 50, 50, 100, 100
    );
    
    EXPECT_EQ(grid_objects.size(), 25);
    
    // Check first circle position
    auto* first = grid_storage.get_circle(grid_objects[0]);
    EXPECT_FLOAT_EQ(first->x, 125); // 100 + 50/2
    EXPECT_FLOAT_EQ(first->y, 125);
    
    // Check last circle position
    auto* last = grid_storage.get_circle(grid_objects[24]);
    EXPECT_FLOAT_EQ(last->x, 325); // 100 + 4*50 + 50/2
    EXPECT_FLOAT_EQ(last->y, 325);
}

TEST_F(BatchOperationsTest, PerformanceBatchTranslate) {
    // Create many objects
    ObjectStorage perf_storage;
    std::vector<ObjectID> many_objects;
    
    const int num_objects = 100000;
    for (int i = 0; i < num_objects; ++i) {
        many_objects.push_back(perf_storage.add_circle(i % 1000, i / 1000, 5));
    }
    
    auto start = std::chrono::high_resolution_clock::now();
    BatchOperations::translate_objects(perf_storage, many_objects, 10, 20);
    auto end = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    double ms = duration.count() / 1000.0;
    
    std::cout << "Translated " << num_objects << " objects in " << ms << " ms\n";
    std::cout << "Performance: " << (num_objects / ms * 1000) << " objects/second\n";
    
    // Should be very fast
    EXPECT_LT(ms, 10); // Less than 10ms for 100k objects
    
    auto stats = BatchOperations::last_operation_stats;
    EXPECT_GT(stats.objects_per_second, 10000000); // > 10M objects/second
}

TEST_F(BatchOperationsTest, PerformanceVsIndividual) {
    const int num_objects = 10000;
    ObjectStorage storage1, storage2;
    std::vector<ObjectID> ids1, ids2;
    
    // Create identical objects in both storages
    for (int i = 0; i < num_objects; ++i) {
        ids1.push_back(storage1.add_circle(i % 100, i / 100, 5));
        ids2.push_back(storage2.add_circle(i % 100, i / 100, 5));
    }
    
    // Time batch operation
    auto start = std::chrono::high_resolution_clock::now();
    BatchOperations::translate_objects(storage1, ids1, 10, 20);
    auto batch_time = std::chrono::high_resolution_clock::now() - start;
    
    // Time individual operations
    start = std::chrono::high_resolution_clock::now();
    for (auto id : ids2) {
        if (auto* circle = storage2.get_circle(id)) {
            circle->x += 10;
            circle->y += 20;
        }
    }
    auto individual_time = std::chrono::high_resolution_clock::now() - start;
    
    auto batch_us = std::chrono::duration_cast<std::chrono::microseconds>(batch_time).count();
    auto individual_us = std::chrono::duration_cast<std::chrono::microseconds>(individual_time).count();
    
    std::cout << "Batch time: " << batch_us << " us\n";
    std::cout << "Individual time: " << individual_us << " us\n";
    
    // Avoid division by zero
    double speedup = (batch_us > 0) ? static_cast<double>(individual_us) / batch_us : 1.0;
    std::cout << "Batch operation speedup: " << speedup << "x\n";
    
    // For small numbers of objects, the overhead of grouping by type
    // might make batch operations slightly slower. The real benefit
    // comes with larger datasets and SIMD operations.
    // Just verify both methods work correctly.
    
    // Verify results are the same
    for (size_t i = 0; i < num_objects; ++i) {
        auto* c1 = storage1.get_circle(ids1[i]);
        auto* c2 = storage2.get_circle(ids2[i]);
        EXPECT_FLOAT_EQ(c1->x, c2->x);
        EXPECT_FLOAT_EQ(c1->y, c2->y);
    }
}

File: tests/cpp_integration/test_batch_operations.py
#!/usr/bin/env python3
"""Test batch operations with C++ drawing library."""

import sys

sys.path.insert(0, "cpp")

import drawing_cpp
from python.drawing_cpp_wrapper import DrawingCpp
import time
import math


def test_batch_operations():
    """Test batch geometric operations."""
    print("Testing Batch Operations")
    print("========================")

    # Create drawing with grid of circles
    drawing = DrawingCpp(1000, 1000)
    storage = drawing._drawing.get_storage()

    # Create a 10x10 grid using batch operation
    grid_ids = drawing_cpp.BatchOperations.create_grid(
        storage,
        drawing_cpp.ObjectType.Circle,
        10,
        10,  # rows, cols
        50,
        50,  # cell width/height
        100,
        100,  # offset
    )

    print(f"Created {len(grid_ids)} circles in a grid")

    # Test batch translate
    print("\nTranslating all objects by (50, 50)...")
    start = time.time()
    drawing_cpp.BatchOperations.translate_objects(storage, grid_ids, 50, 50)
    elapsed = (time.time() - start) * 1000

    stats = drawing_cpp.get_last_operation_stats()
    print(f"  Time: {stats.time_ms:.3f} ms")
    print(f"  Objects/second: {stats.objects_per_second:,.0f}")

    # Test batch scale
    print("\nScaling all objects by 1.5x...")
    center = drawing_cpp.Point(350, 350)
    drawing_cpp.BatchOperations.scale_objects(storage, grid_ids, 1.5, 1.5, center)

    # Test batch rotate
    print("\nRotating all objects by 45 degrees...")
    drawing_cpp.BatchOperations.rotate_objects(storage, grid_ids, math.radians(45), center)

    # Test batch alignment
    print("\nAligning objects to left edge...")
    drawing_cpp.BatchOperations.align_objects_left(storage, grid_ids[:10])

    # Calculate bounding box
    bbox = drawing_cpp.BatchOperations.calculate_bounding_box(storage, grid_ids)
    print(f"\nBounding box of all objects:")
    print(f"  Min: ({bbox.min_x:.1f}, {bbox.min_y:.1f})")
    print(f"  Max: ({bbox.max_x:.1f}, {bbox.max_y:.1f})")
    print(f"  Size: {bbox.width():.1f} x {bbox.height():.1f}")

    # Save result
    drawing.save_json("batch_operations_result.json")
    print("\nSaved result to batch_operations_result.json")


def benchmark_batch_vs_individual():
    """Compare batch operations vs individual operations."""
    print("\n\nBenchmark: Batch vs Individual Operations")
    print("==========================================")

    num_objects = 50000

    # Create two identical drawings
    drawing1 = DrawingCpp(5000, 5000)
    drawing2 = DrawingCpp(5000, 5000)

    # Add many circles
    print(f"Creating {num_objects} circles...")
    ids1 = []
    ids2 = []

    for i in range(num_objects):
        x = (i % 1000) * 5
        y = (i // 1000) * 5
        ids1.append(drawing1.add_circle(x, y, 2))
        ids2.append(drawing2.add_circle(x, y, 2))

    # Time batch operation
    storage1 = drawing1._drawing.get_storage()
    start = time.time()
    drawing_cpp.BatchOperations.translate_objects(storage1, ids1, 100, 100)
    batch_time = time.time() - start

    # Time individual operations
    start = time.time()
    for obj_id in ids2:
        storage2 = drawing2._drawing.get_storage()
        # Note: In real use, we'd have a better API for this
        # For now, we use batch operation with single item
        drawing_cpp.BatchOperations.translate_objects(storage2, [obj_id], 100, 100)
    individual_time = time.time() - start

    print(f"\nResults for {num_objects} objects:")
    print(f"  Batch operation: {batch_time*1000:.1f} ms")
    print(f"  Individual operations: {individual_time*1000:.1f} ms")
    print(f"  Speedup: {individual_time/batch_time:.1f}x")

    # Performance with different operations
    print("\nTesting different batch operations:")

    operations = [
        ("Translate", lambda s, ids: drawing_cpp.BatchOperations.translate_objects(s, ids, 10, 10)),
        ("Scale", lambda s, ids: drawing_cpp.BatchOperations.scale_objects(s, ids, 1.1, 1.1)),
        (
            "Calculate bbox",
            lambda s, ids: drawing_cpp.BatchOperations.calculate_bounding_box(s, ids),
        ),
    ]

    for op_name, op_func in operations:
        start = time.time()
        op_func(storage1, ids1)
        elapsed = (time.time() - start) * 1000

        stats = drawing_cpp.get_last_operation_stats()
        if stats.objects_processed > 0:
            print(
                f"  {op_name}: {elapsed:.2f} ms ({stats.objects_per_second/1e6:.1f}M objects/sec)"
            )


if __name__ == "__main__":
    test_batch_operations()
    benchmark_batch_vs_individual()

